src/main/java/io/jenkins/plugins/sshbuildagents/ssh/mina/ConnectionImpl.java

```diff
@@ -0,0 +1,363 @@
+            // TODO document the permissions the file needs and how to set the umask
+            // TODO verify if the file exists and if the content is the same
+            // TODO verify if the file is a directory
+            permissions.add(PosixFilePermission.OWNER_WRITE);
+            scp.upload(bytes, remotePath, permissions, null);
+        }
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public void setServerHostKeyAlgorithms(String[] algorithms) {}
+
+    /** {@inheritDoc} */
+    @Override
+    public void setTCPNoDelay(boolean tcpNoDelay) {
+        this.tcpNoDelay = tcpNoDelay;
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public ClientSession connect() throws IOException {
+        initClient();
+        if (isSession() == false) {
+            for (int i = 0; i <= maxNumRetries; i++) {
+                try {
+                    return connectAndAuthenticate();
+                } catch (Exception ex) {
+                    String message = getExMessage(ex);
+                    if (maxNumRetries - i > 0) {
+                        println(
+                                stderr,
+                                "SSH Connection failed with IOException: \""
+                                        + message
+                                        + "\", retrying in "
+                                        + retryWaitTime
+                                        + " seconds."
+                                        + " There are "
+                                        + (maxNumRetries - i)
+                                        + " more retries left.");
+                    }
+                }
+                waitToRetry();
+            }
+            throw new IOException("Max number or reties reached.");
+        }
+        return session;
+    }
+
+    /**
+     * @return True is the session is authenticated and open.
+     */
+    private boolean isSession() {
+        return session != null && session.isAuthenticated() && session.isOpen();
+    }
+
+    /**
+     * Connets to the SSH service configured and authenticate
+     *
+     * @return Returns a ClientSession connected and authenticated.
+     * @throws IOException in case of error.
+     * @throws InterruptedException
+     */
+    private ClientSession connectAndAuthenticate() throws IOException, InterruptedException {
+        ConnectFuture connectionFuture = client.connect(this.credentials.getUsername(), this.host, this.port);
+        connectionFuture.verify(this.timeoutMillis);
+        session = connectionFuture.getSession();
+        var authenticator = SSHAuthenticator.newInstance(session, credentials);
+        authenticator.authenticate(TaskListener.NULL);
+        return session;
+    }
+
+    /** Initialize the SSH client. It reuses the client if it exists. */
+    private void initClient() {
+        if (client == null) {
+            client = SshClient.setUpDefaultClient();
```

Comment:
SshClient is meant to be reuse to avoid excessive allocation. 
We can have a single instance, I have implemented this https://github.com/olamy/ssh-agents-plugin/commit/d87a31cc3afd71a5b0d67819c9df84e1371d68c0
saddly you have forced push to your branch which has totally broken the branch and PR to your repo

Reply:
Why? Can you explain your reasoning? 
This is very different from what is recommended by the Apache Mina project itself. 
See javadoc https://javadoc.io/doc/org.apache.sshd/sshd-core/2.14.0/org/apache/sshd/client/SshClient.html
For installation with large number of agents, this will consume/create a lot resources (such threadpool, io etc..) while it's not needed.